Migrate Iceberg catalog (#1692)
redpanda-data/documentation: modules/manage/pages/iceberg/migrate-iceberg-catalog.adoc

= Migrate Iceberg Catalogs
:description: Switch the Iceberg catalog backend for an existing Redpanda cluster without losing untranslated topic data.

// tag::single-source[]
:page-topic-type: how-to
:page-categories: Iceberg, Migration
:personas: ops_admin, streaming_developer
:learning-objective-1: Verify that a target Iceberg catalog supports your existing schemas and partition specs
:learning-objective-2: Pause Iceberg translation and drain pending commits without losing untranslated data
:learning-objective-3: Apply new catalog configuration and resume translation safely

> Switch your cluster from one Iceberg catalog backend to another without losing untranslated topic data. Use this procedure when moving from the filesystem-based `object_storage` catalog to a managed REST catalog, or when changing between REST catalogs.

Reviewer: This first line seems abrupt, like it came out of some other context. The second sentence seems like a better opening sentence.

Use the procedures in this topic when moving from the filesystem-based `object_storage` catalog to a managed REST catalog, or when changing between REST catalogs. By doing so, you can switch your cluster from one Iceberg catalog backend to another without losing untranslated topic data.

> The procedure pauses Iceberg translation per topic, lets pending commits drain to the old catalog, applies the new catalog configuration, and restarts the cluster. While translation is paused, retention is temporarily set to infinite to prevent untranslated data from being deleted.

Reviewer: ```suggestion
This procedure guides you through pausing Iceberg translation per topic, draining pending commits to the old catalog, applying the new catalog configuration, and restarting the cluster.
```

====

== Run the migration

. Save the current `retention.ms` and `retention.bytes` values for every Iceberg topic, then set both to `-1` (infinite retention):
+
[,bash]
----
rpk topic alter-config <topic-name> --set retention.ms=-1 --set retention.bytes=-1
----
+
While Iceberg translation is paused in the next step, the topic's retention anchor on the log is released. Without infinite retention, the cluster could delete untranslated data before the migration completes.

. Pause Iceberg translation on every Iceberg topic by setting `redpanda.iceberg.mode` to `disabled`. Save each topic's previous mode value so you can restore it later.
+
[,bash]
----
rpk topic alter-config <topic-name> --set redpanda.iceberg.mode=disabled
----
+
Setting the mode to `disabled` stops new translation while letting already-translated data finish committing to the old catalog. For more about Iceberg modes, see xref:manage:iceberg/specify-iceberg-schema.adoc[].
+
NOTE: Do not change config_ref:iceberg_enabled,true,properties/cluster-properties[`iceberg_enabled`] at the cluster level. The Iceberg integration must remain enabled at the cluster level so that pending commits can drain to the old catalog.

> . Wait for pending commits to drain. Monitor the `redpanda_iceberg_pending_commit_lag` Prometheus metric on every broker until it reaches `0` for every Iceberg topic-partition.

Author: @wdberkeley  We don't currently have `redpanda_iceberg_pending_commit_lag` documented for Cloud or Self-managed. We do currently have [vectorized_cluster_partition_iceberg_offsets_pending_commit](https://docs.redpanda.com/current/reference/internal-metrics-reference/#vectorized_cluster_partition_iceberg_offsets_pending_commit) and [vectorized_cluster_partition_iceberg_offsets_pending_translation](https://docs.redpanda.com/current/reference/internal-metrics-reference/#vectorized_cluster_partition_iceberg_offsets_pending_translation) documented for Self-managed, under Internal Metrics. 

Is `redpanda_iceberg_pending_commit_lag` a new metric that should be added to the Public Metrics page? (Claude also found `redpanda_iceberg_pending_translation_lag`).

Reviewer 2: Hm good question. In the broker code the metrics are `redpanda_iceberg_pending_translation_lag` and `redpanda_iceberg_pending_commit_lag`. Those other metrics sound like they could be synonyms or aggregates that we expose but I'm not sure. I think that's a question for Cloud. If users don't have access to the pending commit metric, then they can't safely do the migration without support looking at the metric for them.

Author: I found https://github.com/redpanda-data/redpanda/commit/adccab8a84 and I do believe these metrics are also exposed in Cloud -- I'll file a separate doc ticket to look closer into why our metrics doc generation process didn't pull those in.

+

> This metric reports the number of offsets pending a commit to the Iceberg catalog. While it is non-zero, Redpanda is still flushing translated data to the old catalog. The translation lag metric, `redpanda_iceberg_pending_translation_lag`, can remain non-zero. That value reflects new records the cluster has not yet translated, which is expected while translation is paused.

Reviewer: "Which is expected" has an ambiguous antecedent — it could refer to the metric being non-zero, the untranslated records, or the whole situation

> This metric reports the number of offsets pending a commit to the Iceberg catalog. While it is non-zero, Redpanda is still flushing translated data to the old catalog. The translation lag metric, `redpanda_iceberg_pending_translation_lag`, can remain non-zero. That value reflects new records the cluster has not yet translated, which is expected while translation is paused.

Reviewer: ```suggestion
This metric reports the number of offsets pending a commit to the Iceberg catalog. While it is non-zero, Redpanda is still flushing translated data to the old catalog. A non-zero value here is expected while translation is paused, and reflects new records the cluster has not yet translated.
```

+
[TIP]
====
If you scrape Prometheus, the following expression returns `0` only when every Iceberg-topic partition has fully drained:

[,promql]
----
sum(redpanda_iceberg_pending_commit_lag)
----
====

. Apply the new catalog configuration. For example, to switch from `object_storage` to a REST catalog, update the catalog cluster properties:
+
[,bash]
----
rpk cluster config set iceberg_catalog_type rest
rpk cluster config set iceberg_rest_catalog_endpoint <endpoint-url>
rpk cluster config set iceberg_rest_catalog_authentication_mode oauth2
# Set additional credential properties for your chosen authentication mode.
----
+
For full guidance on setting catalog cluster properties, see xref:manage:iceberg/use-iceberg-catalogs.adoc#rest[Connect to a REST catalog] and the individual xref:manage:iceberg/rest-catalog/index.adoc[REST catalog integration pages].

. Restart Redpanda. The catalog cluster properties require a restart to take effect.
+
ifndef::env-cloud[]
For instructions, see xref:manage:cluster-maintenance/rolling-restart.adoc[].
endif::[]
ifdef::env-cloud[]

> Coordinate the restart with Redpanda Support. The restart must occur after `redpanda_iceberg_pending_commit_lag` has reached `0` (step 3) and before you resume translation in the next step.

Reviewer: Can Cloud customers not initiate a rolling restart themselves? If they can, this guidance may be overly restrictive. If they can't, is "Coordinate with Redpanda Support" the right instruction, or is there a specific support process to reference? I see you included @wzzzrd86 here--good--as we should always check with support before placing guidance in the docs to contact support.

Reviewer 3: Approved to direct them to Support here, cloud customers currently cannot kick off a cloud restart

> endif::[]

Author: @wdberkeley could you please confirm if the rolling restart is the correct guidance for Self-managed, and that this restart isn't self-serve for Cloud users?

Reviewer 2: Yes, a rolling restart of the whole cluster started after commit lag has dropped to zero for all topic-partitions and before re-enabling translation is the right move. The restart is required on all brokers so that they initialize with the correct catalog and the correct catalog authentication.

I'm not sure about how Cloud users do restarts... is it really not self-serve for them? I think #help-cloud could confirm quickly.

Author: @wzzzrd86 Could you please confirm that we're good to include this restart guidance for Cloud users in this doc?

. After the cluster comes up, check broker logs for successful catalog requests and the absence of authentication errors to verify the new catalog connection.

. Resume Iceberg translation by restoring `redpanda.iceberg.mode` on every Iceberg topic to its previous value:
+
[,bash]
----
rpk topic alter-config <topic-name> --set redpanda.iceberg.mode=<previous-mode>
----

> . Restore `retention.ms` and `retention.bytes` on every Iceberg topic to the values you saved in step 1.

Reviewer: Referencing a step by number breaks if the order changes. The style guide discourages numbered step references in body text for the same reason it discourages step numbers in headings.

> . Restore `retention.ms` and `retention.bytes` on every Iceberg topic to the values you saved in step 1.

Reviewer: ```suggestion
. Restore `retention.ms` and `retention.bytes` on every Iceberg topic to the values you saved before starting the migration.
```

== Troubleshooting

* Pending commits stuck after restart: A schema or partition spec mismatch between the original tables and the new catalog is the most common cause. See <<verify-catalog-compatibility,Verify catalog compatibility>>. If you cannot resolve the mismatch, contact https://support.redpanda.com/hc/en-us/requests/new[Redpanda Support^].
* Authentication errors against the new REST catalog: Verify that the credential cluster properties (for example, `iceberg_rest_catalog_client_id`, `iceberg_rest_catalog_client_secret`, `iceberg_rest_catalog_token`) match what the new catalog expects. For OAuth, also check `iceberg_rest_catalog_oauth2_server_uri`.
* Translation does not resume after restoring `redpanda.iceberg.mode`: Check that `redpanda_iceberg_pending_translation_lag` is increasing as new records are produced. If it remains `0`, the cluster is not translating new records. Verify that your producer is still writing to the topic and that the topic's mode value is one of `key_value`, `value_schema_id_prefix`, or `value_schema_latest`.

> 

Author: @wdberkeley I also just remembered that we had previously published an upgrade guide for 25.3 https://docs.redpanda.com/current/upgrade/iceberg-schema-changes-and-migration-guide/ Would you mind double checking if there is anything that might be contradictory in these two docs or needs to be taken into account, if both docs happen to apply to any users?

Reviewer 2: No contradictions. If we need to address customers doing both, we should advice they follow the upgrade procedure first, in its entirety, and then switch catalogs.
